### ISSUE TITLE
Executing Custom Code for Domains Not Found in Any List

### ISSUE THREAD

**Issue by eternity4318** [pi-hole/FTL#2092]
I want to run custom code when a domain query is not found on any blocklists. I believe the correct place for this is in [`dnsmasq_interface.c`](https://github.com/pi-hole/FTL/blob/61a211f1c187206f5ff901afae657968114fde15/src/dnsmasq_interface.c#L1205). I tried adding logging there but didn't get any results.

Is there a recommended way to hook into Pi-hole's query handling for this purpose? Any guidance on where to implement this or how to debug would be appreciated.

**Comment by eternity4318** [comment]
I'm trying to run custom code whenever a domain isn't found on any blocklists. To get familiar with the code, I started by adding logging to see what data I have access to and to start hooking into Pi-hole:

```c
if (!blockDomain) {
    // Inline file logging
    FILE *log_file = fopen("/var/log/pihole/logfile.log", "a");
    if (log_file != NULL) {
        fprintf(log_file, "%s\n", name);
        fclose(log_file);
    } else {
        perror("Error opening file"); 
    }
}
```

For some reason, though, nothing is getting logged. Ideally, I’d like to set this up to call a function in C, Bash, or any other language, passing in the relevant data to decide on blocking. If this seems like a feature that could benefit others, I’d be happy to submit a pull request once it’s fully implemented.

**Comment by rdwebdesign** [comment]
If your intention is to identify if a domain is blocked by Pi-hole, you could try to use `curl` in a shell script (or any other language) to call this API endpoint:

```
https://pi.hole/api/search/<DOMAIN>?partial=false
```

and check the returned JSON.

If `results.total == 0`, the domain is not present in any list or a regex:
```json
{
   "results": {
      "domains":{ "exact":0, "regex":0 },
      "gravity":{ "allow":0, "block":0 },
      "total":0
   },
   "parameters": { "N":20, "partial":false, "domain":"example.com", "debug":false }
}
```

**Note:**
The API call needs authentication.

**Comment by eternity4318** [comment]
I want it to be done as the request is being processed. So that I have full control over what's returned as a response.

**Comment by eternity4318** [comment]
Any update?

**Comment by DL6ER** [comment]
No, I don't think there is anything w are supposed to do right now. Do you have a specific question we can answer? The code snippet you posted above lacks context, e.g., where precisely you have tried adding it. Without this, it's unclear what you want. Offloading into a separate (external) environment is not necessarily a trivial task.
